Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp

```diff
@@ -530,69 +530,198 @@ const char* TerrainSystem::GetMaxSurfaceName(
+    }
+}
+
+void TerrainSystem::ProcessSurfaceWeightsFromList(
+    const AZStd::vector<AZ::Vector3>& inPositions,
+    AzFramework::Terrain::SurfaceWeightsListFillCallback perPositionCallback,
+    Sampler sampleFilter) const
+{
+    if (!perPositionCallback)
+    {
+        return;
+    }
+
+    for (const auto& position : inPositions)
+    {
+        bool terrainExists = false;
+        AzFramework::SurfaceData::SurfaceTagWeightList surfaceWeights;
+        GetSurfaceWeights(position, surfaceWeights, sampleFilter, &terrainExists);
+        perPositionCallback(position, surfaceWeights, terrainExists);
+    }
+}
+
+void TerrainSystem::ProcessSurfacePointsFromList(
+    const AZStd::vector<AZ::Vector3>& inPositions,
+    AzFramework::Terrain::SurfacePointListFillCallback perPositionCallback,
+    Sampler sampleFilter) const
+{
+    if (!perPositionCallback)
+    {
+        return;
+    }
 
-void TerrainSystem::ProcessHeightsFromRegion(const AZ::Aabb& inRegion, const AZ::Vector2 stepSize, Sampler sampleFilter, SurfacePointRegionFillCallback perPositionCallback, TerrainDataReadyCallback onComplete)
+    for (const auto& position : inPositions)
+    {
+        bool terrainExists = false;
+        AzFramework::SurfaceData::SurfacePoint surfacePoint;
+        GetSurfacePoint(position, surfacePoint, sampleFilter, &terrainExists);
+        perPositionCallback(position, surfacePoint, terrainExists);
+    }
+}
+
+
+void TerrainSystem::ProcessHeightsFromRegion(
+    const AZ::Aabb& inRegion,
+    const AZ::Vector2& stepSize,
+    AzFramework::Terrain::HeightRegionFillCallback perPositionCallback,
+    Sampler sampleFilter) const
 {
     // Don't bother processing if we don't have a callback
     if (!perPositionCallback)
     {
         return;
     }
 
-    uint32_t numSamplesX = static_cast<uint32_t>((inRegion.GetMax().GetX() - inRegion.GetMin().GetX()) / stepSize.GetX());
-    uint32_t numSamplesY = static_cast<uint32_t>((inRegion.GetMax().GetY() - inRegion.GetMin().GetY()) / stepSize.GetY());
+    const size_t numSamplesX = aznumeric_cast<size_t>(ceil(inRegion.GetExtents().GetX() / stepSize.GetX()));
+    const size_t numSamplesY = aznumeric_cast<size_t>(ceil(inRegion.GetExtents().GetY() / stepSize.GetY()));
 
-    for (uint32_t y = 0; y < numSamplesY; y++)
+    for (size_t y = 0; y < numSamplesY; y++)
     {
-        for (uint32_t x = 0; x < numSamplesX; x++)
+        float fy = (float)(inRegion.GetMin().GetY() + (y * stepSize.GetY()));
+        for (size_t x = 0; x < numSamplesX; x++)
         {
+            bool terrainExists = false;
             float fx = (float)(inRegion.GetMin().GetX() + (x * stepSize.GetX()));
-            float fy = (float)(inRegion.GetMin().GetY() + (y * stepSize.GetY()));
-
-            SurfaceData::SurfacePoint surfacePoint;
-            GetHeight(AZ::Vector3(fx, fy, 0.0f), sampleFilter, surfacePoint.m_position);
-            perPositionCallback(surfacePoint, x, y);
+            AZ::Vector3 inPosition(fx, fy, 0.0f);
```

Comment:
You can construct inPosition once outside the loop then just update its fx / fy.

Reply:
Fixed